api/webhook/validate_auth.go

```diff
@@ -0,0 +1,165 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	authv1 "k8s.io/api/authorization/v1"
+	authorizationv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+)
+
+var authLog = ctrl.Log.WithName("validate-auth")
+
+// +kubebuilder:webhook:path=/validate-auth,mutating=false,failurePolicy=fail,groups=chaos-mesh.org,resources=*,verbs=create;update,versions=v1alpha1,name=vauth.kb.io
+
+// AuthValidator validates the authority
+type AuthValidator struct {
+	enabled bool
+	client  client.Client
+	reader  client.Reader
+	authCli *authorizationv1.AuthorizationV1Client
+
+	decoder *admission.Decoder
+
+	clusterScoped     bool
+	targetNamespace   string
+	allowedNamespaces string
+	ignoredNamespaces string
+}
+
+// NewAuthValidator returns a new AuthValidator
+func NewAuthValidator(enabled bool, client client.Client, reader client.Reader, authCli *authorizationv1.AuthorizationV1Client,
+	clusterScoped bool, targetNamespace, allowedNamespaces, ignoredNamespaces string) *AuthValidator {
+	return &AuthValidator{
+		enabled:           enabled,
+		client:            client,
+		reader:            reader,
+		authCli:           authCli,
+		clusterScoped:     clusterScoped,
+		targetNamespace:   targetNamespace,
+		allowedNamespaces: allowedNamespaces,
+		ignoredNamespaces: ignoredNamespaces,
+	}
+}
+
+// AuthValidator admits a pod iff a specific annotation exists.
+func (v *AuthValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if !v.enabled {
+		return admission.Allowed("")
+	}
+
+	username := req.UserInfo.Username
+	groups := req.UserInfo.Groups
+	chaosKind := req.Kind.Kind
+
+	// these chaos doesn't contain selector field
+	if chaosKind == v1alpha1.KindAwsChaos || chaosKind == v1alpha1.KindPodNetworkChaos || chaosKind == v1alpha1.KindPodIoChaos {
+		return admission.Allowed("")
+	}
+
+	chaos := v1alpha1.GetChaosValidator(chaosKind)
+	if chaos == nil {
+		err := fmt.Errorf("kind %s is not support", chaosKind)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err := v.decoder.Decode(req, chaos)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	specs := chaos.GetSelectSpec()
+
+	requireClusterPrivileges := false
+	affectedNamespaces := make(map[string]struct{})
+
+	for _, spec := range specs {
+		if len(spec.GetSelector().Namespaces) == 0 {
+			requireClusterPrivileges = true
```

Comment:
We can use `continue` directly here! In addition, I didn't understand why we need to set `requireClusterPrivileges` to `true` when `len(spec.GetSelector().Namespaces)` is 0.

Reply:
If we use `Pods` field directly, which includes namespace information, Chaos Mesh will ignore `Namespaces` field. In this case, we still can't limit the permissions of selectors.

Reply:
update in 7202ec8